dll/win32/shell32/dialogs/drive.cpp

```diff
@@ -47,18 +59,86 @@ typedef struct
+                       NULL, OPEN_EXISTING, 0, NULL);
+}
 
-    if (!GetEnvironmentVariableW(L"SystemDrive", wszSystemDrv, _countof(wszSystemDrv)))
-        return FALSE;
+static BOOL
+GetLogicalDriveSize(WORD DriveNumber, ULARGE_INTEGER &Result)
+{
+    BOOL bSuccess = FALSE;
+    const WCHAR szDrivePath[] = { WCHAR(DriveNumber + 'A'), ':', '\\', '\0' };
 
-    if (!_wcsicmp(wszDriveLetter, wszSystemDrv))
-        return TRUE;
+    HANDLE hDevice = OpenLogicalDriveHandle(DriveNumber);
+    if (hDevice != INVALID_HANDLE_VALUE)
+    {
+        GET_LENGTH_INFORMATION LengthInfo;
+        bSuccess = DevIoCtl(hDevice, IOCTL_DISK_GET_LENGTH_INFO, NULL, 0, &LengthInfo, sizeof(LengthInfo));
+        Result.QuadPart = LengthInfo.Length.QuadPart;
+        if (!bSuccess && GetDriveTypeW(szDrivePath) == DRIVE_REMOVABLE) // Blank floppy
+        {
+            DISK_GEOMETRY dg;
+            bSuccess = DevIoCtl(hDevice, IOCTL_DISK_GET_DRIVE_GEOMETRY, NULL, 0, &dg, sizeof(dg));
+            if (bSuccess)
+                Result.QuadPart = dg.Cylinders.QuadPart * dg.TracksPerCylinder * dg.SectorsPerTrack * dg.BytesPerSector;
+        }
+        CloseHandle(hDevice);
+    }
+    if (!bSuccess || !Result.QuadPart)
+        bSuccess = GetDiskFreeSpaceExW(szDrivePath, NULL, &Result, NULL); // Note: Not exact if NTFS quotas are in effect
+    return bSuccess;
+}
 
-    return FALSE;
+static PWSTR
+GetTipText(FORMAT_DRIVE_CONTEXT &Ctx)
+{
+    HANDLE hDevice = OpenLogicalDriveHandle(Ctx.Drive);
+    if (hDevice != INVALID_HANDLE_VALUE)
+    {
+        ULONG cb;
+        ZeroMemory(Ctx.szTip, sizeof(Ctx.szTip));
+        UNICODE_STRING &NameInfo = (UNICODE_STRING&)Ctx.szTip;
+        NtQueryObject(hDevice, ObjectNameInformation, Ctx.szTip, sizeof(Ctx.szTip), &cb);
+        if (!NameInfo.Buffer || NameInfo.Buffer[0] != '\\')
+            (NameInfo.Buffer = Ctx.szTip)[0] = UNICODE_NULL;
+
+        PARTITION_INFORMATION_EX pie;
+        pie.PartitionStyle = PARTITION_STYLE_RAW;
+        if (!DevIoCtl(hDevice, IOCTL_DISK_GET_PARTITION_INFO_EX, NULL, 0, &pie, sizeof(pie)))
+        {
+            PARTITION_INFORMATION pi;
+            if (DevIoCtl(hDevice, IOCTL_DISK_GET_PARTITION_INFO, NULL, 0, &pi, sizeof(pi)))
+            {
+                pie.PartitionStyle = PARTITION_STYLE_MBR;
+                pie.PartitionNumber = pi.PartitionNumber;
+            }
+        }
+        WCHAR buf[150], guidbuf[39];
+        buf[0] = UNICODE_NULL;
+        if (pie.PartitionStyle == PARTITION_STYLE_GPT)
+        {
+            StringFromGUID2(pie.Gpt.PartitionId, guidbuf, _countof(guidbuf));
+            StringCchPrintfW(buf, _countof(buf), L"\nGPT %s %s", guidbuf, pie.Gpt.Name);
+        }
+        if (pie.PartitionStyle == PARTITION_STYLE_MBR)
+        {
+            StringCchPrintfW(buf, _countof(buf), L"\nMBR (%d)", pie.PartitionNumber);
+        }
+        StringCbCatW(NameInfo.Buffer, sizeof(Ctx.szTip) - sizeof(UNICODE_STRING), buf);
+        CloseHandle(hDevice);
```

Comment:
You could move this up just after the IOCTL_DISK_GET_PARTITION_INFO calls above.